pangeo_forge_runner/plugin.py

```diff
@@ -0,0 +1,147 @@
+"""
+Handle the plugin system for injections.
+
+There are three parts of injections:
+
+1. An "injection spec", provided by other installed packages (such as
+   pangeo_forge_recipes, pangeo_forge_cmr, etc). This specifies what
+   *values* exactly will be injected as args for *which* callables.
+   It is in the form of a dictionary, and looks like this:
+   ```
+   {
+     "<callable-1-name>": {
+       "<argument-1-name>": "<value-spec>",
+       "<argument-2-name>": "<value-spec>"
+     },
+     "<callable-2-name>": {
+       "<argument-1-name>": "<value-spec>",
+       "<argument-2-name>": "<value-spec>"
+     }
+   }
+   ```
+
+   `<value-spec>` specifies what value should be injected. Currently
+   supported are two strings:
+   1. `OUTPUT_ROOT` - Root path that *output* should be written to. Will
+      be an FSSpec path.
+   2. `CACHE_ROOT` - (Optional) Root path that should be used for caching
+      input values if necessary.
+
+   Additional values may be provided in the future.
+
+   An example is:
+
+   ```
+    {
+        'StoreToZarr': {
+            'target_root': 'OUTPUT_ROOT',
```

Comment:
@yuvipanda, I'm gradually understanding this. Thanks again for your patience as I review it, it's a great learning opportunity for me.

Question: Did you have a rationale for not going 1:1 on naming here, i.e.

```diff
         'StoreToZarr': {
-            'target_root': 'OUTPUT_ROOT',
+            'target_root': 'TARGET_ROOT',
```
?

Perhaps you thought `OUTPUT_ROOT` was more general/reusable?

Reply:
@cisaacstern yes, `OUTPUT_ROOT` is more general was the reason. 'target' felt very zarr specific.

Reply:
@yuvipanda I see that, and also the more I reflect on it, the more I find myself caught on the fact that, IIUC, the `OUTPUT_ROOT` is always parsed as:

https://github.com/pangeo-forge/pangeo-forge-runner/blob/8d61651c3b46c538b8216e95d7a1c46a03095a13/pangeo_forge_runner/commands/bake.py#L178-L180

which means it always becomes a `pangeo_forge_recipes.storage.FSSpecTarget` (via `pangeo_forge_runner.storage.TargetStorage`)...

https://github.com/pangeo-forge/pangeo-forge-runner/blob/8d61651c3b46c538b8216e95d7a1c46a03095a13/pangeo_forge_runner/storage.py#L55-L59

https://github.com/pangeo-forge/pangeo-forge-runner/blob/8d61651c3b46c538b8216e95d7a1c46a03095a13/pangeo_forge_runner/storage.py#L75-L80

with that in mind, I think it would be more intuitive to me if this value-spec was called either `FSSPEC_TARGET` after the `-recipes` type it is parsed to (but that's probably too specific) or, after the `-runner` type that parses it `TARGET_STORAGE` (maybe this is better). And the same for the cache target. So to make the suggestion more specific, how about:

```diff
`<value-spec>` specifies what value should be injected. Currently
   supported are two strings:
-   1. `OUTPUT_ROOT` - Root path that *output* should be written to. Will
-      be an FSSpec path.
+   1. `TARGET_STORAGE` - Location *output* should be written to. Will
+      be parsed to an `pangeo_forge_recipes.storge.FSSpecTarget` object.
-   2. `CACHE_ROOT` - (Optional) Root path that should be used for caching
-      input values if necessary.
+   2. `INPUT_CACHE_STORAGE` - (Optional) Location that should be used for caching
+      input values. Parsed to `pangeo_forge_recipes.storge.CacheFSSpecTarget`.
```

This way the injection spec is not inventing its own terminology: we just use the name of the `pangeo-forge-runner` parsing interface as the name of the value-spec. Thoughts?

Reply:
Yeah, I'm totally fine with standardizing and not picking own terminology! I think the question is one of - should we include or not include the 'root' terminology here? Prior to pangeo-forge-runner, 'root' didn't really exist as terminology, but it now does as we put multiple job outputs from the same run - #68 introduced this I think. So the question now is - is this `TARGET_STORAGE` or `TARGET_STORAGE_ROOT`?

Reply:
Awesome.

> So the question now is - is this TARGET_STORAGE or TARGET_STORAGE_ROOT?

I am less attached to one or the other of these choices, than I am to the idea that whichever it is, the value-spec name should be the screaming snake case version of the `-runner` class which will parses it, i.e. (just because I was curious how to do that... 😆):
```python
import re
cls_name = "TargetStorage"
re.sub("(?<!^)(?=[A-Z])", "_", cls_name).upper()
# --> 'TARGET_STORAGE'
```

So I'd say 'TARGET_STORAGE' because:

1. It changes less, because it doesn't require changing the parsing class names (If we prefer 'TARGET_STORAGE_ROOT', then I'd say let's rename `TargetStorage` -> `TargetStorageRoot`.)
2. It's more concise.
3. In terms of its meaning, I think it can be taken to imply "target storage _device_/_location_" in the general sense, and not the fully-qualified final output path.

Thoughts?